Added the rest of the Uri/UriBuilder properties

bjarkihall opened the pull request:
Uri/UriBuilder wrappers only exposed Uri.GetScheme and UriBuilder.GetQuery/SetQuery.
This PR makes sure that both of them have access to all of the properties that make up a URI, that is; 
Scheme, Host, Port, Path, Query and Fragment.

For reference:
[scheme]://[host]:[port][path[?query][#fragment]
[http]://[www.contoso.com]:[80][/catalog/shownew.htm][?date=today][#search]

JesperSchulz commented:
Sounds reasonable :-) We'll take a look! Thanks for submitting!

mazhelez commented on Modules/System Tests/URI/src/UriBuilderTest.Codeunit.al:
```diff
@@ -22,6 +22,83 @@ codeunit 135071 "Uri Builder Test"
+        Uri: Codeunit Uri;
+    begin
+        // [Given] A URI with an http scheme
+        UriBuilder.Init('http://microsoft.com');
```
Maybe assert the right scheme is in place after initialization.

`Assert.AreEqual('https', UriBuilder.Scheme(), 'The scheme does not match');`

mazhelez commented on Modules/System Tests/URI/src/UriBuilderTest.Codeunit.al:
```diff
@@ -22,6 +22,83 @@ codeunit 135071 "Uri Builder Test"
+        UriBuilder.Init('http://microsoft.com');
+
+        // [When] Setting the scheme to https
+        UriBuilder.SetScheme(UriBuilder.GetScheme() + 's');
```
Be explicit here: `UriBuilder.SetScheme('https');`

mazhelez commented on Modules/System Tests/URI/src/UriBuilderTest.Codeunit.al:
```diff
@@ -22,6 +22,83 @@ codeunit 135071 "Uri Builder Test"
+        Uri: Codeunit Uri;
+    begin
+        // [Given] A Url
+        UriBuilder.Init(http://microsoft.com/test);
```
Quotes missing, this won't compile.

mazhelez commented on Modules/System Tests/URI/src/UriBuilderTest.Codeunit.al:
```diff
@@ -22,6 +22,83 @@ codeunit 135071 "Uri Builder Test"
+
+        // [Then] The URI is as expected
+        UriBuilder.GetUri(Uri);
+        Assert.AreEqual(http://www.example.org/test, Uri.GetAbsoluteUri(), 'The host does not match');
```
Quotes missing here too.

mazhelez commented on Modules/System Tests/URI/src/UriBuilderTest.Codeunit.al:
```diff
@@ -22,6 +22,83 @@ codeunit 135071 "Uri Builder Test"
 
+    [Test]
+    [Scope('OnPrem')]
+    procedure SetAndGetSchemeTest()
```
Add a test for setting an incorrect scheme. See [here](https://docs.microsoft.com/en-us/dotnet/api/system.uribuilder.scheme?view=net-5.0#System_UriBuilder_Scheme).

You may even introduce an enum to wrap the "allowed" scheme values.

mazhelez commented on Modules/System Tests/URI/src/UriBuilderTest.Codeunit.al:
```diff
@@ -22,6 +22,83 @@ codeunit 135071 "Uri Builder Test"
+
+    [Test]
+    [Scope('OnPrem')]
+    procedure SetAndGetPortTest()
```
A test for setting an incorrect port, please.

mazhelez reviewed (changes requested):
Overall looks great :)

Tests need to be fixed and consider making the scheme an enum.

bjarkihall commented:
Hi and thanks for the feedback!

I tried out the scheme property for UriBuilder and it appears the [documentation](https://docs.microsoft.com/en-us/dotnet/api/system.uribuilder.scheme?view=net-5.0#System_UriBuilder_Scheme) is a bit misleading, since a user can e.g. clear the scheme by providing an empty-string and might even provide something like:
"invalidscheme://microsoft.com", by doing uriBuilder.Scheme = "invalidscheme"; which throws no error.
What appears to be invalid are schemes with characters like space in them, like "invalid scheme", so I at least added a test case for that and also ports over and under the valid number range.

I also added the Enum, but I don't know of how much value it'll be. There already was a test in UriTest.Codeunit.al since Scheme getter was already supported for that wrapper, would you want me to modify that one to use the enum too?

Btw. before I submitted the PR initially AL Extension for vscode got frozen so I had to disable it (along with the intellisense, which made the missing quotations slip by, sorry about that. I've addressed these too. :)

mazhelez commented:
> I also added the Enum, but I don't know of how much value it'll be. There already was a test in UriTest.Codeunit.al since Scheme getter was already supported for that wrapper, would you want me to modify that one to use the enum too?

I see. My idea was mainly to try limit the possible values of the URI scheme used from the BC application. For instance, it is super fishy if ftp scheme is used.

mazhelez commented:
Hey @bjarkihall 

I ran the tests and there are some failures. Can you take a look, please.

```
Failure - Codeunit Uri Builder Test -  Duration 0.123 seconds
   Success - Test method: InitInvalidURITest - Duration 0.007 seconds)
   Failure - Test method: SetAndGetSchemeTest - Duration 0.027 seconds
      Error:
         Assert.AreEqual failed. Expected:<https://microsoft.com> (Text). Actual:<https://microsoft.com:80/> (Text). The scheme does not match.
      Call Stack:
         Library Assert(CodeUnit 130002).AreEqual - Library Assert by Microsoft
         Uri Builder Test(CodeUnit 135071).SetAndGetSchemeTest - System Application Test by Microsoft
         Test Runner - Mgt(CodeUnit 130454).RunTests - Test Runner by Microsoft
         Test Runner - Isol. Codeunit(CodeUnit 130450).OnRun - Test Runner by Microsoft
         Test Suite Mgt.(CodeUnit 130456).RunTests - Test Runner by Microsoft
         Test Suite Mgt.(CodeUnit 130456).RunSelectedTests - Test Runner by Microsoft
         Test Suite Mgt.(CodeUnit 130456).RunNextTest - Test Runner by Microsoft
         Command Line Test Tool(Page 130455).RunNextTest - OnAction - Test Runner by Microsoft
   Failure - Test method: SetIncorrectPortTest - Duration 0.01 seconds
      Error:
         Assert.ExpectedError failed. Expected: A call to System.UriBuilder.Port failed with this message: Specified argument was out of the range of valid values.\Parameter name: value. Actual: A call to System.UriBuilder.Port failed with this message: Specified argument was out of the range of valid values.\Parameter name: value.
      Call Stack:
         Library Assert(CodeUnit 130002).ExpectedError - Library Assert by Microsoft
         Uri Builder Test(CodeUnit 135071).SetIncorrectPortTest - System Application Test by Microsoft
         Test Runner - Mgt(CodeUnit 130454).RunTests - Test Runner by Microsoft
         Test Runner - Isol. Codeunit(CodeUnit 130450).OnRun - Test Runner by Microsoft
         Test Suite Mgt.(CodeUnit 130456).RunTests - Test Runner by Microsoft
         Test Suite Mgt.(CodeUnit 130456).RunSelectedTests - Test Runner by Microsoft
         Test Suite Mgt.(CodeUnit 130456).RunNextTest - Test Runner by Microsoft
         Command Line Test Tool(Page 130455).RunNextTest - OnAction - Test Runner by Microsoft
         
   Failure - Test method: SetAndGetPathTest - Duration 0.007 seconds
      Error:
         Assert.AreEqual failed. Expected:</segment2/segment3> (Text). Actual:</segment2/segment3/> (Text). GetPath does not work as expected.
      Call Stack:
         Library Assert(CodeUnit 130002).AreEqual - Library Assert by Microsoft
         Uri Builder Test(CodeUnit 135071).SetAndGetPathTest - System Application Test by Microsoft
         Test Runner - Mgt(CodeUnit 130454).RunTests - Test Runner by Microsoft
         Test Runner - Isol. Codeunit(CodeUnit 130450).OnRun - Test Runner by Microsoft
         Test Suite Mgt.(CodeUnit 130456).RunTests - Test Runner by Microsoft
         Test Suite Mgt.(CodeUnit 130456).RunSelectedTests - Test Runner by Microsoft
         Test Suite Mgt.(CodeUnit 130456).RunNextTest - Test Runner by Microsoft
         Command Line Test Tool(Page 130455).RunNextTest - OnAction - Test Runner by Microsoft
```

mazhelez reviewed (changes requested):
Tests are failing and need to be fixed.

bjarkihall commented:
These were a bit weird, since I used the output of GetLastErrorText and the \-newline character didn't seem to work in strict string comparisons but \n\r actually said they were equal. I also didn't think the default port would be part of AbsoluteUri (according to the [docs](https://docs.microsoft.com/en-us/dotnet/api/system.uri.absoluteuri?view=net-5.0#examples)) but I guess UriBuilder adds it?

Oh well, at least they're all green now:
![image](https://user-images.githubusercontent.com/8615374/107460615-54407c00-6b50-11eb-8384-e21622c8f82a.png)

JesperSchulz commented:
Availability update: We will publish a fix for this issue in the next update for release 17.

JesperSchulz commented:
Thanks for reporting this. We agree, and we’ll publish a fix asap, either in an update for the current version or in the next major release. We will update this issue with information about availability. Please do not reply to this, as we do not monitor closed issues. If you have follow-up questions or requests, please create a new issue where you reference this one.
